README.md
#### Table of contents

1. [Description - What the module does and why it is useful](#module-description)
2. [Setup](#setup)
   * [Requirements](#requirements)
   * [Get Azure credentials](#get-azure-credentials)
   * [Installing the Azure module](#installing-the-azure-module)
3. [Usage - Configuration options and additional functionality](#usage)
4. [Reference - An under-the-hood peek at what the module is doing and how](#reference)
   * [Types](#types)
   * [Parameters](#parameters)
5. [Known issues](#known-issues)
6. [Limitations - OS compatibility, etc.](#limitations)
7. [Development - reporting issues and getting support](#development)

## Description

Microsoft Azure exposes a powerful API for creating and managing its Infrastructure as a Service platform. The azure module allows you to drive that API using Puppet code. This allows you to use Puppet to create, stop, restart, and destroy Virtual Machines, and eventually to manage other resources, meaning you can manage even more of your infrastructure as code.

## Setup

### Requirements

*   Ruby Gems as follows (see [Installing the Azure module](#installing-the-azure-module), below).
    *   [azure](https://rubygems.org/gems/azure) 0.7.x
    *   [azure_mgmt_storage](https://rubygems.org/gems/azure_mgmt_storage) 0.14.x
    *   [azure_mgmt_compute](https://rubygems.org/gems/azure_mgmt_compute) 0.14.x
    *   [azure_mgmt_resources](https://rubygems.org/gems/azure_mgmt_resources) 0.14.x
    *   [azure_mgmt_network](https://rubygems.org/gems/azure_mgmt_network) 0.14.x
    *   [hocon](https://rubygems.org/gems/hocon) 1.1.x
*   Azure credentials (as detailed below).

#### Get Azure credentials

To use this module, you need an Azure account. If you already have one, you can skip this section.

First, sign up for an [Azure account](https://azure.microsoft.com/en-us/free/).

Install [the Azure CLI 1.0](https://docs.microsoft.com/en-us/azure/cli-install-nodejs), which is a cross-platform node.js-based tool that works on Windows and Linux. This is required to generate a certificate for the Puppet module, but it's also a useful way of interacting with Azure. Currently these instructions have not been updated for the CLI 2.0 tool ('az' commands) but this module uses the API.

[Register the CLI](https://azure.microsoft.com/en-gb/documentation/articles/xplat-cli-connect/) with your Azure account.

On the command line, enter:

``` shell
azure account download
azure account import <path to your .publishsettings file>
```

After you've created the account, export the PEM certificate file using the following command:

``` shell
azure account cert export
```

Next, get a subscription ID using the `azure account list` command:

``` shell
$ azure account list
info:    Executing command account list
data:    Name                    Id                                     Tenant Id  Current
data:    ----------------------  -------------------------------------  ---------  -------
data:    Pay-As-You-Go           xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx  undefined  true
info:    account list command OK
```

To use the Resource Manager API instead, you need a service principal on the Active Directory. A quick way to create one for Puppet is [pendrica/azure-credentials](https://github.com/pendrica/azure-credentials). Its [puppet mode](https://github.com/pendrica/azure-credentials#puppet-style-output-note--v-displays-the-file-on-screen-after-creation) can create the `azure.conf` (see below) for you. Alternatively, the official documentation covers [creating this and retrieving the required credentials](https://azure.microsoft.com/en-us/documentation/articles/resource-group-authenticate-service-principal/).

### Installing the Azure module

Install the required gems with this command on `puppet-agent` 1.2 (included in Puppet Enterprise 2015.2.0) or later:

``` shell
/opt/puppetlabs/puppet/bin/gem install retries --no-ri --no-rdoc
/opt/puppetlabs/puppet/bin/gem install azure --version='~>0.7.0' --no-ri --no-rdoc
/opt/puppetlabs/puppet/bin/gem install azure_mgmt_compute --version='~>0.14.0' --no-ri --no-rdoc
/opt/puppetlabs/puppet/bin/gem install azure_mgmt_storage --version='~>0.14.0' --no-ri --no-rdoc
/opt/puppetlabs/puppet/bin/gem install azure_mgmt_resources --version='~>0.14.0' --no-ri --no-rdoc
/opt/puppetlabs/puppet/bin/gem install azure_mgmt_network --version='~>0.14.0' --no-ri --no-rdoc
/opt/puppetlabs/puppet/bin/gem install hocon --version='~>1.1.2' --no-ri --no-rdoc
```

When installing on Windows, launch `Start Command Prompt with Puppet` and enter:

``` shell
gem install retries --no-ri --no-rdoc
gem install azure --version="~>0.7.0" --no-ri --no-rdoc
gem install azure_mgmt_compute --version="~>0.14.0" --no-ri --no-rdoc
gem install azure_mgmt_storage --version="~>0.14.0" --no-ri --no-rdoc
gem install azure_mgmt_resources --version="~>0.14.0" --no-ri --no-rdoc
gem install azure_mgmt_network --version="~>0.14.0" --no-ri --no-rdoc
gem install hocon --version="~>1.1.2" --no-ri --no-rdoc
```

On versions of `puppet agent` older than 1.2 (Puppet Enterprise 2015.2.0), use the older path to the `gem` binary:

``` shell
/opt/puppet/bin/gem install retries --no-ri --no-rdoc
/opt/puppet/bin/gem install azure --version='~>0.7.0' --no-ri --no-rdoc
/opt/puppet/bin/gem install azure_mgmt_compute --version='~>0.14.0' --no-ri --no-rdoc
/opt/puppet/bin/gem install azure_mgmt_storage --version='~>0.14.0' --no-ri --no-rdoc
/opt/puppet/bin/gem install azure_mgmt_resources --version='~>0.14.0' --no-ri --no-rdoc
/opt/puppet/bin/gem install azure_mgmt_network --version='~>0.14.0' --no-ri --no-rdoc
/opt/puppet/bin/gem install hocon --version='~>1.1.2' --no-ri --no-rdoc
```

> **Note:** You must pin Azure gem installs to the correct version detailed in the example above for the azure module to work properly. The example above pins the hocon gem version to prevent possible incompatibilities.

Set the following environment variables specific to your Azure installation.

If using the classic API, provide this information:

``` shell
export AZURE_MANAGEMENT_CERTIFICATE='/path/to/pem/file'
export AZURE_SUBSCRIPTION_ID='your-subscription-id'
```

At a Windows command prompt, specify the information **without quotes around any of the values**:


``` shell
SET AZURE_MANAGEMENT_CERTIFICATE=C:\Path\To\file.pem
SET AZURE_SUBSCRIPTION_ID=your-subscription-id
```

If using the Resource Management API, provide this information:

``` shell
export AZURE_SUBSCRIPTION_ID='your-subscription-id'
export AZURE_TENANT_ID='your-tenant-id'
export AZURE_CLIENT_ID='your-client-id'
export AZURE_CLIENT_SECRET='your-client-secret'
```

At a Windows command prompt, specify the information **without quotes around any of the values**:

``` shell
SET AZURE_SUBSCRIPTION_ID=your-subscription-id
SET AZURE_TENANT_ID=your-tenant-id
SET AZURE_CLIENT_ID=your-client-id
SET AZURE_CLIENT_SECRET=your-client-secret
```

If you are working with **both** Resource Manager and classic virtual machines, provide all of the above credentials.

Alternatively, you can provide the information in a configuration file of [HOCON format](https://github.com/typesafehub/config). Store this as `azure.conf` in the relevant [confdir](https://docs.puppetlabs.com/puppet/latest/reference/dirs_confdir.html):

* \*nix Systems: `/etc/puppetlabs/puppet`
* Windows: `C:\ProgramData\PuppetLabs\puppet\etc`
* Non-root users: `~/.puppetlabs/etc/puppet`

The file format is:

``` shell
azure: {
  subscription_id: "your-subscription-id"
  management_certificate: "/path/to/pem/file"
}
```

When creating this file on Windows, note that as a JSON-based config file format, paths must be properly escaped:

``` shell
azure: {
  subscription_id: "your-subscription-id"
  management_certificate: "C:\\path\\to\\file.pem"
}
```

> **Note**: Make sure to have at least hocon 1.1.2 installed on windows. With older versions, you have to make sure to make sure that the `azure.conf` is encoded as UTF-8 without a byte order mark (BOM). See [HC-82](https://tickets.puppetlabs.com/browse/HC-82), and [HC-83](https://tickets.puppetlabs.com/browse/HC-83) for technical details. Starting with hocon 1.1.2, UTF-8 with or without BOM works.

Or, with the Resource Management API:

``` shell
azure: {
  subscription_id: "your-subscription-id"
  tenant_id: "your-tenant-id"
  client_id: "your-client-id"
  client_secret: "your-client-secret"
}
```

You can use either the environment variables **or** the config file. If both are present, the environment variables are used. You cannot have some settings in environment variables and others in the config file.

Next, Install the module with:

``` shell
puppet module install puppetlabs-azure
```

## Usage

### Create Azure VMs

Azure has two modes for deployment: Classic and Resource Manager. For more information, see [Azure Resource Manager vs. classic deployment: Understand deployment models and the state of your resources](https://azure.microsoft.com/en-us/documentation/articles/resource-manager-deployment-model/). The module supports creating VMs in both deployment modes.

#### Classic

You can create Azure classic virtual machines using the following:

```puppet
azure_vm_classic { 'virtual-machine-name':
  ensure           => present,
  image            => 'b39f27a8b8c64d52b05eac6a62ebad85__Ubuntu-14_04_2-LTS-amd64-server-20150706-en-us-30GB',
  location         => 'West US',
  user             => 'username',
  size             => 'Medium',
  private_key_file => '/path/to/private/key',
}
```

#### Resource Manager

You can create Azure Resource Manager virtual machines using the following:

```puppet
azure_vm { 'sample':
  ensure         => present,
  location       => 'eastus',
  image          => 'canonical:ubuntuserver:14.04.2-LTS:latest',
  user           => 'azureuser',
  password       => 'Password_!',
  size           => 'Standard_A0',
  resource_group => 'testresacc01',
}
```

You can also add a [virtual machine extension](https://azure.microsoft.com/en-us/documentation/articles/virtual-machines-windows-extensions-features/) to the VM and deploy from a [Marketplace product](https://azure.microsoft.com/en-us/blog/working-with-marketplace-images-on-azure-resource-manager/) instead of an image:

```puppet
azure_vm { 'sample':
  ensure         => present,
  location       => 'eastus',
  user           => 'azureuser',
  password       => 'Password_!',
  size           => 'Standard_A0',
  resource_group => 'testresacc01',
  plan           => {
    'name'      => '2016-1',
    'product'   => 'puppet-enterprise',
    'publisher' => 'puppet',
  },
  extensions     => {
    'CustomScriptForLinux' => {
       'auto_upgrade_minor_version' => false,
       'publisher'                  => 'Microsoft.OSTCExtensions',
       'type'                       => 'CustomScriptForLinux',
       'type_handler_version'       => '1.4',
       'settings'                   => {
         'commandToExecute' => 'sh script.sh',
         'fileUris'         => ['https://myAzureStorageAccount.blob.core.windows.net/pathToScript']
       },
     },
  },
}
```

This type also has many other properties you can manage:

```puppet
azure_vm { 'sample':
  location                      => 'eastus',
  image                         => 'canonical:ubuntuserver:14.04.2-LTS:latest',
  user                          => 'azureuser',
  password                      => 'Password',
  size                          => 'Standard_A0',
  resource_group                => 'testresacc01',
  storage_account               => 'teststoracc01',
  storage_account_type          => 'Standard_GRS',
  os_disk_name                  => 'osdisk01',
  os_disk_caching               => 'ReadWrite',
  os_disk_create_option         => 'fromImage',
  os_disk_vhd_container_name    => 'conttest1',
  os_disk_vhd_name              => 'vhdtest1',
  dns_domain_name               => 'mydomain01',
  dns_servers                   => '10.1.1.1.1 10.1.2.4',
  public_ip_allocation_method   => 'Dynamic',
  public_ip_address_name        => 'ip_name_test01pubip',
  virtual_network_name          => 'vnettest01',
  virtual_network_address_space => '10.0.0.0/16',
  subnet_name                   => 'subnet111',
  subnet_address_prefix         => '10.0.2.0/24',
  ip_configuration_name         => 'ip_config_test01',
  private_ip_allocation_method  => 'Dynamic',
  network_interface_name        => 'nicspec01',
  network_security_group_name   => 'My-Network-Security-Group',
  tags                          => { 'department' => 'devops', 'foo' => 'bar' },
  extensions                    => {
    'CustomScriptForLinux' => {
       'auto_upgrade_minor_version' => false,
       'publisher'                  => 'Microsoft.OSTCExtensions',
       'type'                       => 'CustomScriptForLinux',
       'type_handler_version'       => '1.4',
       'settings'                   => {
         'commandToExecute' => 'sh script.sh',
         'fileUris'         => ['https://myAzureStorageAccount.blob.core.windows.net/pathToScript']
       },
     },
  },
}
```

#### Premium Storage

Azure supports _premium_ SSD backed VMs for enhanced performance of production class environments.  SSD storage can be selected at the time of VM creation like this (`Premium_LRS` is the Azure API's internal representation):

```puppet
azure_vm { 'ssd-example':
  ensure               => present,
  location             => 'centralus',
  image                => 'Canonical:UbuntuServer:16.10:latest',
  user                 => 'azureuser',
  password             => 'Password_!',
  size                 => 'Standard_DS1_v2',
  resource_group       => 'puppetvms',
  storage_account_type => 'Premium_LRS',
}

```

To successfully enable `Premium_LRS`, you **must** select a premium-capable VM size such as `Standard_DS1_v2`.  Regular HDD backed VMs can be created by using `Standard_LRS`.

#### Boot/guest diagnostics

The Azure portal provides switches to enable _boot_diagnostics_ and _guest diagnostics_.  Both switches require access to a storage account to dump the diagnostic data.

The switch behaves differently depending what is activated:

* Boot diagnostics - Configures the VM `diagnosticsProfile` setting to write out boot diagnostics .  If required, manually enable using the portal.  Since boot diagnostics only apply at boot time, their most useful for interactive debugging when a VM is having a problems booting.  If required, boot diagnostics can be enabled through the Azure portal.
* Guest diagnostics - Configures an extension to capture live diagnostic output.  This needs to be _different_ depending on the selected guest OS and is enabled by supplying the appropriate data to the `extensions` parameter.

#### Managed Disks

Azure's _managed disks_ feature removes the requirement to associate a storage account with each Azure VM. To use managed disks with `azure_vm`, set the `managed_disks` parameter to true:

```puppet
azure_vm { 'managed-disks-example':
  ensure        => present,
  location      => 'centralus',
  image         => 'Canonical:UbuntuServer:16.10:latest',
  user          => 'azureuser',
  password      => 'Password_!',
  managed_disks => true,
}
```

When using _managed disks_ it's not possible to set _vhd_ options, the _managed disks_ feature takes care of these for you.

#### Connecting to networks

You can create Azure Resource Manager virtual networks using the following:

```puppet
azure_virtual_network { 'vnettest01':
  ensure           => present,
  location         => 'eastus',
  address_prefixes => ['10.0.0.0/16'], # Array of IP address prefixes for the VNet
  dns_servers      => [],              # Array of DNS server IP addresses
}
```

Specify the network objects to avoid creating your VM in a miniture DMZ where it can't reach other networks. To attach a VM to a virtual network, specify the `virtual_network_name`, `subnet_name` and `network_security_group_name` parameters. These all allow slashes to lookup the requested object in other resource groups.  Note that `subnet_name` must also specify the virtual network if using this feature:

```puppet
azure_vm { 'web01':
  ensure                      => present,
  location                    => 'centralus',
  image                       => 'canonical:ubuntuserver:14.04.2-LTS:latest',
  user                        => 'azureuser',
  password                    => 'Password_!',
  size                        => 'Standard_A0',
  resource_group              => 'webservers-rg',
  virtual_network_name        => 'hq-rg/delivery-vn',
  subnet_name 	              => "hq-rg/delivery-vn/web-sn",
  network_security_group_name => "hq-rg/delivery-nsg",
}
```

If virtual network parameters specified in the `azure_vm` do not exist, they will be created in the same resource group as the VM.  This works for basic environments where everything you want to talk to on non-public addresses is within the same resource group. You can avoid this automatic creation by not specifying `virtual_network_address_space`

```puppet
azure_vm { 'web01':
  ensure                        => present,
  location                      => 'centralus',
  resource_group                => 'webservers-rg',
  virtual_network_name          => 'vnettest01',
  virtual_network_address_space => '10.0.0.0/16',
  ...
}
```

### List and manage VMs

This module supports listing and managing machines via `puppet resource`.

For example:

``` shell
puppet resource azure_vm_classic
```

This outputs some information about the machines in your account:

```puppet
azure_vm_classic { 'virtual-machine-name':
  ensure        => 'present',
  cloud_service => 'cloud-service-uptjy',
  deployment    => 'cloud-service-uptjy',
  hostname      => 'garethr',
  image         => 'b39f27a8b8c64d52b05eac6a62ebad85__Ubuntu-14_04_2-LTS-amd64-server-20150706-en-us-30GB',
  ipaddress     => 'xxx.xx.xxx.xx',
  location      => 'West US',
  media_link    => 'http://xxx.blob.core.windows.net/vhds/disk_2015_08_28_07_49_34_868.vhd',
  os_type       => 'Linux',
  size          => 'Medium',
}
```

Use the same command for Azure Resource Manager:

``` shell
puppet resource azure_vm
```

This lists Azure Resource Manager VMs:

```puppet
azure_vm { 'sample':
  location       => 'eastus',
  image          => 'canonical:ubuntuserver:14.04.2-LTS:latest',
  user           => 'azureuser',
  password       => 'Password',
  size           => 'Standard_A0',
  resource_group => 'testresacc01',
}
```

### Create Azure storage accounts

You can create a [storage account](https://azure.microsoft.com/en-us/documentation/articles/storage-create-storage-account/) using the following:

```puppet
azure_storage_account { 'myStorageAccount':
  ensure         => present,
  resource_group => 'testresacc01',
  location       => 'eastus',
  account_type   => 'Standard_GRS',
}
```

> **Note:** Storage accounts are created with the Azure Resource Manager API only.

### Create Azure resource groups

You can create a [resource group](https://azure.microsoft.com/en-us/documentation/articles/resource-group-overview/#resource-groups) using the following:

```puppet
azure_resource_group { 'testresacc01':
  ensure   => present,
  location => 'eastus',
}
```

> **Note:** Resource groups are created with Azure Resource Manager API only.

### Create Azure template deployment

You can create a [resource template deployment](https://azure.microsoft.com/en-us/documentation/articles/solution-dev-test-environments/) using the following:

```puppet
azure_resource_template { 'My-Network-Security-Group':
  ensure         => 'present',
  resource_group => 'security-testing',
  source         => 'https://gallery.azure.com/artifact/20151001/Microsoft.NetworkSecurityGroup.1.0.0/DeploymentTemplates/NetworkSecurityGroup.json',
  params         => {
    'location'                 => 'eastasia',
    'networkSecurityGroupName' => 'testing',
  },
}
```

> **Note:** Resource templates are deployed with Azure Resource Manager API only.

## Reference

### Types

* [`azure_vm_classic`](#type-azure_vm_classic): Manages a virtual machine in Microsoft Azure with Classic Service Management API.
* `azure_vm`: Manages a virtual machine in Microsoft Azure with Azure Resource Manager API.
* `azure_storage_account`: Manages a Storage Account with Azure Resource Manager API.
* `azure_resource_group`: Manages a Resource Group with Azure Resource Manager API.
* `azure_resource_template`: Manages a Resource Template with Azure Resource Manager API.

### Parameters


CHANGELOG.md
## Supported Version 1.3.1
## Added
 - Japanese translations updated to reflect the 1.3.0 README.md.

## Supported Version 1.3.0
## Added
 - New Puppet Types and Providers for subnet, vnet and network_security_group.
 - Support for the Microsoft ARM 0.14 API
 - Support for SKU in Storage Accounts.
 - Types for access_tier and https_traffic_only in Storage Accounts.
 - Overhaul of acceptance testing to run locally on the host.

## Fixed
 - Acceptance tests now can run locally on your host. (bundle exec rake fast) This should complete in ~ 24 mins and verify the basic stack on ARM. To commit PR's please continue to run (bundle exec rake acceptance).

## Supported Version 1.2.1
### Added
 - This release adds internationalization for the README.md.
 - The first supported language is Japanese.

## Supported Version 1.2.0
### Added
 - Support for Azure's managed disks feature removes the requirement to associate a storage account with each Azure VM, removing one of the fundamental limitations of the platform.
 - Premium SSD backed VMs for enhanced performance of production class environments are now supported and documented.
 - The network security group can now be specified when creating a VM.
 - Cross networking support. By default, all network objects created while provisioning a azure_vm will be created in the resource group you created the VM in. This works fine in basic environments where everything you want to talk to on non-public addresses is within the same resource group but if you need to plug into a network in another resource group.
 - Tagging is now supported for Puppet Types.

### Fixed
 - Redundant acceptance tests removed.

## Supported Version 1.1.1
This release adds support for Puppet 4.9.4.

### Added
- Puppet 4.9.4 Support

### Fixed
- `azure_resource_template` The templateURL parameter when unused was being
set to Nil and failing.
- `ProviderArm` parameter `virtual_network_address_space` now supports an array of
addresses


## Supported Version 1.1.0
### Summary
This release adds several more useful resource types for managing Azure resource groups, storage accounts, and resource templates. It also expands the capabilities of the `azure_vm` type by adding support for managing extensions, data disks, custom data, and deploying from marketplace images.

### Added
- `azure_resource_group` type for resource group management
- `azure_storage_account` type for storage account management
- `azure_resource_template` type for template management
- `extensions` parameter to `azure_vm` for extension configuration
- `plan` parameter to `azure_vm` for Azure Marketplace images
- `data_disks` parameter to `azure_vm` for data disk configuration
- `custom_data` parameter to `azure_vm` for custom data configuration

### Fixed
- Updated to 1.1.2 version of hocon gem (includes windows fixes)
- No longer requires azure.conf classic credentials for only ARM or vice versa
- Better printing of error messages from azure's API
- Allow `azure_vm::public_ip_allocation_method => 'None'` to work


## Supported Version 1.0.3
### Summary
This release updates the module for the 0.3.0 version of the azure gems and
fixes a bug for Azure::Core

### Fixed
- Update to 0.3.0 version of azure\_mgmt\* gems
- Fix Azure::Core require
- Fix paging of REST when lots of VMs
- Fix puppet resource failing validation
- Fix docs mentioning incorrect quoting in azure.conf
- Fix bundlered listen gem failing on older rubies
- Fix lint warnings in examples/\*.pp
- Add Debian 8 to metadata
- Fix domain to be spec-specific

## Supported Version 1.0.2

This release includes:
* (CLOUD-488) Windows agent support for the Azure module testing.
* Several test improvements and fixes.
* Fixes validation for name length.
* Updates module for Hocon 1.0.0 compatibility.
* Improves error reporting.
* Adds apt-get update before install for custom_data param.

## 2015-12-08 - Supported Version 1.0.1

This release includes:

* Updates to the metadata to identify supported Windows platforms
* Improvements to the documentation around Windows support
* Fixes to URLs pointing at the previous Git repository

## 2015-12-08 - Supported Release 1.0.0

The first public release of the Azure module provides support for
managing VMs in both the Service Management (Classic) and new Resource
Management APIs in Azure.


lib/puppet/type/azure_vm.rb
require 'puppet/parameter/boolean'

require_relative '../../puppet_x/puppetlabs/azure/property/read_only'
require_relative '../../puppet_x/puppetlabs/azure/property/positive_integer'
require_relative '../../puppet_x/puppetlabs/azure/property/string'
require_relative '../../puppet_x/puppetlabs/azure/property/hash'
require_relative '../../puppet_x/puppetlabs/azure/property/boolean'

# azure_vm { 'sample':
#   location         => 'West US'
#   image            => 'canonical:ubuntuserver:14.04.2-LTS:latest',
#   user             => 'azureuser',
#   password         => 'Password',
#   size             => 'Standard_A0',
#   resource_group   => 'myresourcegroup',
# }

Puppet::Type.newtype(:azure_vm) do
  @doc = 'Type representing a virtual machine in Microsoft Azure.'

  validate do
    required_properties = [
      :location,
    ]
    required_properties.each do |property|
      # We check for both places so as to cover the puppet resource path as well
      if self[:ensure] == :present and self[property].nil? and self.provider.send(property) == :absent
        fail "You must provide a #{property}"
      end
    end

    managed_disks_conflicts = [
      :storage_account,
      :os_disk_name,
      :os_disk_vhd_name,
      :os_disk_vhd_container_name,
    ]
    if self[:managed_disks]
      managed_disks_conflicts.each do |property|
        if ! self[property].nil?
          fail "You cannot provide #{property} when manged_disks is set to true"
        end
      end
    end
  end

  newproperty(:ensure) do
    defaultto :present
    newvalue(:present) do
      provider.create unless provider.exists?
    end
    newvalue(:absent) do
      provider.destroy if provider.exists?
    end
    newvalue(:running) do
      if provider.exists?
        provider.start unless provider.running?
      else
        provider.create
      end
    end
    newvalue(:stopped) do
      if provider.exists?
        provider.stop unless provider.stopped?
      else
        provider.create
        provider.stop
      end
    end
    def change_to_s(current, desired)
      current = :running if current == :present
      desired = current if desired == :present and current != :absent
      current == desired ? current : "changed #{current} to #{desired}"
    end
    def insync?(is)
      is.to_s == should.to_s or
        (is.to_s == 'running' and should.to_s == 'present') or
        (is.to_s == 'stopped' and should.to_s == 'present')
    end
  end

  newparam(:name, namevar: true, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'Name of the virtual machine.'
    validate do |value|
      super value
      fail("the name must be between 1 and 64 characters long") if value.size > 64
    end
  end

  newproperty(:image, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'Name of the image to use to create the virtual machine.'
    validate do |value|
      super value
      fail("the image name must not be empty") if value.empty?
    end
    def insync?(is)
      is.casecmp(should).zero?
    end
  end

  newproperty(:user, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'User name for the virtual machine. This value is only used when creating the VM initially.'
  end

  newparam(:password, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The password for the virtual machine. This value is only used when creating the VM initially.'
    validate do |value|
      super value
      fail 'the password must not be empty' if value.empty?
    end
  end

  newproperty(:location, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The location where the virtual machine will be created.'
    validate do |value|
      super value
      fail 'the location must not be empty' if value.empty?
    end
  end

  newproperty(:size, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The size of the virtual machine instance.'
    validate do |value|
      super value
      fail 'the size must not be empty' if value.empty?
    end
  end

  newproperty(:resource_group, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The name of the associated resource group'
    validate do |value|
      super value
      fail 'the resource group must not be empty' if value.empty?
      fail("the resource group must be less that 65 characters in length") if value.size > 64
    end
    def insync?(is)
      is.casecmp(should).zero?
    end
  end

  newparam(:storage_account, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The name of the associated storage account'
    validate do |value|
      super value
      fail 'the storage account must not be empty' if value.empty?
    end
  end

  newproperty(:extensions) do
    desc 'The hash of extensions and their settings'
    def insync?(is)
      should.reduce(true) do |insync,(name,should_props)|
        if is[name]
          # Filter properties that are nil or unmanaged
          is_props = is[name].reject { |k,v| v.nil? or k == 'provisioning_state' }
          # Propagate any mismatches
          insync && (is_props == should_props)
        else
          # Skip any extensions not defined to be managed
          true
        end
      end
    end
  end

  newproperty(:data_disks) do
    desc 'Enables the configuration one or more data disks on a VM.

These disks have a number of required properties:
- name
- lun
- caching
- disk_size_gb (for new)
- vhd (uri)
- create_option
'
    validate do |value|
      value.each do |name,props|
        # these three are required and are freeform
        fail "#{name}: disk_size_gb is required" if props['disk_size_gb'].nil?
        fail "#{name}: disk_size_gb must not be empty" if props['disk_size_gb'].empty?
        fail "#{name}: vhd is required" if props['vhd'].nil?
        fail "#{name}: vhd must not be empty" if props['vhd'].empty?
        fail "#{name}: lun is required" if props['lun'].nil?
        fail "#{name}: lun must not be empty" if props['lun'].empty?
        # caching is not required
        if ! props['caching'].nil?
          fail "#{name}: caching must not be empty" if props['caching'].empty?
          unless ['None','ReadOnly','ReadWrite'].include?(props['caching'])
            fail "#{name}: }caching must be one of None, ReadOnly, ReadWrite; got #{props['caching']}"
          end
        end
        # create_option defaults to Empty
        if props['create_option'].nil?
          props['create_option'] = 'Empty'
        end
        fail "#{name}: create_option must not be empty" if props['create_option'].empty?
        unless ['FromImage','Empty','Attach'].include?(props['create_option'])
          fail "#{name}: }create_option must be one of FromImage, Empty, Attach; got #{props['create_option']}"
        end
      end
    end
    munge do |value|
      value.each do |name, props|
        props['disk_size_gb'] = props['disk_size_gb'].to_i if props['disk_size_gb']
        props['lun'] = props['lun'].to_i if props['lun']
      end
    end
    def insync?(is)
      should.reduce(true) do |insync, (name,should_props)|
        if is[name]
          # Filter properties that are nil or unmanaged
          is_props = is[name].reject { |k,v| v.nil? or should[name][k].nil? }
          # Propagate any mismatches
          insync && (is_props == should_props)
        else
          # Skip any extensions not defined to be managed
          true
        end
      end
    end
  end

  newparam(:custom_data, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'A script to be executed on launch by Cloud-Init. Linux guests only.'
  end

  newparam(:storage_account_type, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The name of the associated storage account type'
    defaultto 'Standard_GRS'
  end

  newproperty(:os_disk_name, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The name of the associated os disk name'
  end

  newproperty(:os_disk_caching, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The disk cache setting'
    defaultto 'ReadWrite'
  end

  newproperty(:os_disk_create_option, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The os disk create option'
    defaultto 'FromImage'
  end

  newproperty(:os_disk_vhd_container_name, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The os disk vhd container name'
    defaultto do
      if ! resource[:managed_disks]
        'vhds'
      end
    end
  end

  newproperty(:os_disk_vhd_name, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The os disk vhd name'
  end

  newproperty(:plan, :parent => PuppetX::PuppetLabs::Azure::Property::Hash) do
    desc 'Hash of properties for plan.

The keys that should be in the hash are:
- name
- product
- publisher
- promotion_code (optional)'
  end

  newparam(:public_ip_address_name, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The public ip address name'
  end

  newparam(:dns_domain_name, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The dns domain name for the vm'
  end

  newparam(:dns_servers, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The dns servers for the vm'
    defaultto '10.1.1.1 10.1.2.4'
  end

  newparam(:public_ip_allocation_method, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The public ip allocation method'
    defaultto 'Dynamic'
  end

  newparam(:virtual_network_name, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The virtual network name'
  end

  newparam(:virtual_network_address_space, :array_accepting => :all) do
    desc 'The virtual network address space'
    defaultto '10.0.0.0/16'
  end

  newparam(:subnet_name, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The subnet name'
    defaultto 'default'
  end

  newparam(:subnet_address_prefix, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The subnet address prefix'
    defaultto '10.0.2.0/24'
  end

  newparam(:ip_configuration_name, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The ip configuration name'
  end

  newparam(:private_ip_allocation_method, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The private ip allocation method'
    defaultto 'Dynamic'
  end

  newproperty(:network_interface_name, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The network interface name'
  end

  newproperty(:managed_disks, :parent => PuppetX::PuppetLabs::Azure::Property::Boolean) do
    desc 'Use Azure\'s managed disks feature'

    defaultto false
  end

  newproperty(:network_security_group_name, :parent => PuppetX::PuppetLabs::Azure::Property::String) do
    desc 'The network security group name'
    def insync?(is)
      true
    end
  end

  newproperty(:tags, :parent => PuppetX::PuppetLabs::Azure::Property::Hash) do
    desc 'The tags for the instance'
    defaultto {}
  end

  autorequire(:azure_resource_group) do
    self[:resource_group]
  end
end
